File: packages/core-dart/lib/src/address/parse.dart
import 'detect.dart';
import 'codes.dart';

ParseResult parse(String input) {
  final kind = detect(input);
  if (kind == AddressKind.invalid) {
    return ParseResult(
      kind: kind,
      address: input,
      warnings: [],
      error: AddressError(
        code: ErrorCode.unknownPrefix,
        input: input,
        message: 'Invalid address',
      ),
    );
  }

  final warnings = <Warning>[];
  if (input != input.toUpperCase()) {
    warnings.add(
      Warning(
        code: WarningCode.nonCanonicalAddress,
        severity: 'warn',
        message: 'Address normalized to uppercase',
        normalization: Normalization(
          original: input,
          normalized: input.toUpperCase(),
        ),
      ),
    );
  }

  return ParseResult(
    kind: kind,
    address: input.toUpperCase(),
    warnings: warnings,
  );
}

File: packages/core-go/address/parse.go
package address

import "strings"

func Parse(input string) ParseResult {
	kind := Detect(input)
	if kind == "invalid" {
		return ParseResult{Kind: "invalid", Err: &AddressError{Code: ErrUnknownPrefix, Input: input, Message: "Invalid address"}}
	}

	warnings := []Warning{}
	if input != strings.ToUpper(input) {
		warnings = append(warnings, Warning{
			Code:     WarnNonCanonicalAddress,
			Severity: "warn",
			Message:  "Address normalized to uppercase",
			Normalization: &Normalization{Original: input, Normalized: strings.ToUpper(input)},
		})
	}

	return ParseResult{Kind: kind, Address: strings.ToUpper(input), Warnings: warnings}
}
